modules/blockexplorer/addblock.go

```diff
@@ -0,0 +1,390 @@
+// instead of siacoins
+func appendNewSFOutput(additions []persist.BoltItem, outputID types.SiafundOutputID, txid crypto.Hash) []persist.BoltItem {
+	additions = appendHashType(additions, crypto.Hash(outputID), hashFundOutputID)
+	return append(additions, persist.BoltItem{
+		BucketName: "SiafundOutputs",
+		Key:        encoding.Marshal(outputID),
+		Value: encoding.Marshal(outputTransactions{
+			OutputTx: txid,
+		}),
+	})
+}
+
+func appendHashType(changes []persist.BoltItem, hash crypto.Hash, hashType int) []persist.BoltItem {
+	return append(changes, persist.BoltItem{
+		BucketName: "Hashes",
+		Key:        encoding.Marshal(hash),
+		Value:      encoding.Marshal(hashType),
+	})
+}
+
+// Parses and adds
+func (be *BlockExplorer) addBlockDB(b types.Block) error {
+	// Special case for the genesis block, which does not have a
+	// valid parent, and for testing, as tests will not always use
+	// blocks in consensus
+	var blocktarget types.Target
+	if b.ID() == be.genesisBlockID {
+		blocktarget = types.RootDepth
+	} else {
+		var exists bool
+		blocktarget, exists = be.cs.ChildTarget(b.ParentID)
+		if build.DEBUG {
+			if build.Release == "testing" {
+				blocktarget = types.RootDepth
+			}
+			if !exists {
+				panic("Applied block not in consensus")
+			}
+
+		}
+	}
+
+	// Create an additions slice for this block
+	modifications := make([]persist.BoltModification, 0)
+	additions := make([]persist.BoltItem, 0)
+
+	// Construct the struct that will be inside the database
+	blockStruct := blockData{
+		Block:  b,
+		Height: be.blockchainHeight,
+	}
+
+	additions = append(additions, persist.BoltItem{
+		BucketName: "Blocks",
+		Key:        encoding.Marshal(b.ID()),
+		Value:      encoding.Marshal(blockStruct),
+	})
+
+	bSum := blockSummary{
+		ID:        b.ID(),
+		Timestamp: b.Timestamp,
+		Target:    blocktarget,
+		Size:      uint64(len(encoding.Marshal(b))),
+	}
+
+	additions = append(additions, persist.BoltItem{
+		BucketName: "Heights",
+		Key:        encoding.Marshal(be.blockchainHeight),
+		Value:      encoding.Marshal(bSum),
+	})
+
+	additions = appendHashType(additions, crypto.Hash(b.ID()), hashBlock)
+
+	// Add the block to the Transactions bucket so that lookups on
+	// the block as a txid result in a special case, not an
+	// error. Necessary because of miner payouts
+	additions = append(additions, persist.BoltItem{
+		BucketName: "Transactions",
+		Key:        encoding.Marshal(b.ID()),
+		Value:      encoding.Marshal(txInfo{b.ID(), -1}),
+	})
+	// TODO add a dummy txid with block id and txNum of -1, so
+	// that txid lookups on a block id are still valid
+
+	// Insert the miner payouts as new outputs
+	for i := range b.MinerPayouts {
+		modifications = appendAddress(modifications, b.MinerPayouts[i].UnlockHash, crypto.Hash(b.ID()))
+		additions = appendHashType(additions, crypto.Hash(b.MinerPayouts[i].UnlockHash), hashUnlockHash)
+		additions = appendNewOutput(additions, b.MinerPayoutID(i), crypto.Hash(b.ID()))
```

Comment:
instead of performing the append inside the functions, do it here:
```go
modifications = append(modifications, newAddressDiff(b.MinerPayouts[i].UnlockHash, crypto.Hash(b.ID()))
```
this makes functions easier to test, and is generally more "functional" than passing around a pointer to be modified.